### ISSUE TITLE
Using in python scripts

### ISSUE THREAD

**Issue by Phlya** [deeptools/pyGenomeTracks#8]
Hi!

The plots look really nice, I was just wondering whether this library can be used in python scripts to plot on matplotlib axes, and not as a standalone CLI tool?

Thanks!
Ilya

**Comment by bgruening** [comment]
@Phlya I'm not sure what you mean with plotting axes, but you can use pyGenomeTracks as a library from your own scripts.

**Comment by Phlya** [comment]
@bgruening cool, thanks - is there an example somewhere I could follow?

**Comment by bgruening** [comment]
The deepTools API documentation will get you some examples. We need to write some for pyGenomeTracks, but maybe this already get you started.
```
bag@bag:~/projects/code/tools-iuc/tools$ conda create -n pygenometracks pygenometracks ipython -y
bag@bag:~/projects/code/tools-iuc/tools$ source activate pygenometracks
(pygenometracks) bag@bag:~/projects/code/tools-iuc/tools$ ipython
Python 3.5.4 | packaged by conda-forge | (default, Dec  9 2017, 16:18:50) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

In [9]: from pygenometracks import readBed

In [10]: foo = readBed.ReadBed(open('/home/bag/projects/code/galaxy/test-data/2.bed'))

In [11]: foo.fields
Out[11]: 
['chromosome',
 'start',
 'end',
 'name',
 'score',
 'strand',
 'thick_start',
 'thick_end',
 'rgb',
 'block_count',
 'block_sizes',
 'block_starts']

```

**Comment by Phlya** [comment]
And then that object has a `.plot` method, is that what you mean?

**Comment by fidelram** [comment]
take a look at https://github.com/deeptools/pyGenomeTracks/blob/master/pygenometracks/plotTracks.py

In summary you may want to do:

```
import pygenometracks.tracksClass
trp = pygenometracks.tracksClass.PlotTracks(tracks_file_name, fig_width)
trp.plot(output_file_name, region)

```

**Comment by Phlya** [comment]
OK, thanks! What I was thinking is to use it directly without a configuration file, to set up my own matplotlib figure, and plot on it using `pyGenomeTracks`. Is it possible?

**Comment by fidelram** [comment]
Actually yes. For example if you want to plot a bigwig file you can do:

```python
import matplotlib.pyplot as plt
from pyGenomeTracks import BigWigTrack

#make a properties dict using as keys the values from the 
# documentation
properties_dict = {'file': <path to file>, 'height': 3, 'title'='bigwig', 'color'='red}

# initialize a bigwig track with the properties_dict. Any of the available tracks can be
# initialized in the same way
bw = BigWigTrack(properties_dict)

# to plot the region two axes are needed, one for the plot itself and the other for the legend.
figure, (track_ax, label_ax) = plt.subplots(1, 2)

# the plot method of all track classes expects 5 values, the track_ax, 
# the label_ax, and the genomic region to plot (chrom, start, end).
bw.plot(track_ax, label_ax, chrom, start, end)
```

You can do that for any other track class (e.g. BedTrack, LinksTrack, HiCMatrixTrack, etc).

Let me know if it works, maybe you can add an example to the README.md in case someone else has the same issue.

**Comment by Phlya** [comment]
Brilliant, thank you! Had to spend some time to figure some minor things out, but I can make beautiful plots now, like this:
![image](https://user-images.githubusercontent.com/2895034/37358246-8329a4a2-26e2-11e8-9abf-1a9c1b8cb532.png)
The only minor concerns for me are
1) the requirement to specify `label_ax` - here I just used the same axes on the right for all three columns and plotted them on top of each other; I would suggest to make it optional
2) automatic y-range labelling, which might overlap and hide the data - not sure what would be the best approach, but I think it should be optional too, at least

**Comment by Phlya** [comment]
Also, why does using `ax` as `label_ax` produce strange results with the label somewhere very far away from the rest of the plot?

**Comment by fidelram** [comment]
It is very nice that it worked for you.

Honestly, I didn't consider the case in which someone would like to use the track classes independently, that is why there is some hard coded behaviour with respect to the label_ax.

At some point I thought about making the labex_ax independent of the .plot functions but for the Hi-C case, in the label_ax, the colorbar is plotted.  But you are right, I could make it optional.

What happens in the label_ax is that whatever is on the properties_dict['title'] is plotted there. Thus, you can indeed use ax, as label_ax and simply do not put anything for the title. 

The reason why the title is put far away is because the label_ax coordinates are expected to be from 0 to 1 for x an y and the text is plot at y=0.5 and for x I think the position is 0.2. In other coordinate system this will produce unexpected results.

With respect to the y-range labeling, you can turn it off by setting: properties_dict['show data range'] = 'no'.

If you want to look at all the options for a track you can take a look at the the .OPTIONS_TXT constant of each track class that you want to use.

**Comment by Phlya** [comment]
I see, thank you, that's very helpful!

Concerning the title, looking at the code it should be positioned using the fractional axes coordinates (`label_ax.transAxes`), so I thought a different data range shouldn't be a problem...

**Comment by fidelram** [comment]
I think I am not using the `label_ax.transAxes` consistently in all track classes. But you are right, if it is used you should get the 'title' to be positioned towards the left of the ax. I have noticed some inconsistencies with the expected behavior and what I get and maybe what you are seeing is one of those cases. Let me know if you figure out what is wrong.

**Comment by Phlya** [comment]
You are right, I am using `BigWigTrack`, and there the coordinates are absolute. I would suggest making sure they are all fractional, this would be consistent and would make life much easier...

**Comment by Phlya** [comment]
Cool, thank you!

**Comment by iamjli** [comment]
> Brilliant, thank you! Had to spend some time to figure some minor things out, but I can make beautiful plots now, like this:
> ![image](https://user-images.githubusercontent.com/2895034/37359280-14e02fd6-26e5-11e8-89c4-228826d3c291.png)
> 
> The only minor concerns for me are
> 
> 1. the requirement to specify `label_ax` - here I just used the same axes on the right for all three columns and plotted them on top of each other; I would suggest to make it optional
> 2. automatic y-range labelling, which might overlap and hide the data - not sure what would be the best approach, but I think it should be optional too, at least

@Phlya Would you mind sharing some template code you used to generate this figure? I just spent a very frustrating few hours with matplotlib and made very little progress.

**Comment by Phlya** [comment]
Hi Johnny,

I am on holidays now, I can send it to you after Tuesday, but best if you
remind me then in case I forget.

Ilya

**Comment by Phlya** [comment]
Hi @iamjli,

Here is a gist with some code that plots smth very similar to the figure above: https://gist.github.com/Phlya/21b73a7900ab9637d8dafc69dc7537a7

It's quite lengthy, but I have very particular stylistic preferences :) Let me know if you have any problems.

**Comment by iamjli** [comment]
Thanks @Phlya , I'll give it a try!

**Comment by mooratov** [comment]
hello,

So I take it that this is perhaps not the intended use case for this library, but i am also interested in using the track classes in jupyter notebooks / python scripts. 

As i posted in the (closed) jupyter notebook issue, i can make it work if i do something like this

```
from pygenometracks import tracks
properties_dict = {'file': 'whatever.bed', 'height': 3, 'title':'dong', 'color':'red'}
bed = tracks.BedTrack(properties_dict)

figure = plt.figure(figsize=(10,10))
figure, axes = plt.subplots(1,1)
axes.set_xlim(12123100, 12125000) 
bed.plot(axes, 'chr14', 12123120, 12125000)
bed.savefig('test.pdf')
```
but the  result of the above is fairly ugly and not at all the beautiful plots made by the executables. Wondering if there's a chance the developers could accomodate stuff like this. 

Thanks!

**Comment by fidelram** [comment]
Do you mind posting an example? Your code looks correct.

**Comment by shouldsee** [comment]
Can we have one of such example included in `README.md` please?

**Comment by sebgra** [comment]
Hello !

I tried to work with pygenometracks  through python scripts. I used  : 
`from pygenometracks import tracks
`

to import the library in a conda env with : 
` _pygenometracks            3.0                        py_0    bioconda_`

Nevertheless I am stuck in importing the library and I got ModuleImportError : No module named 'pygenometracks'.

I also tried the camel case typing for import the library but it's the same. 

Any solution ? 

Thanks in advance
